### PR TITLE
libudev: several cleanups for udev-list

### LINKED ISSUE #10854: [Regression] #10787 prevents system from booting
**systemd version the issue has been seen with**
> Current git master HEAD

**Used distribution**
 > Exherbo

**Expected behaviour you didn't see**
 > System boots as usual

**Unexpected behaviour you saw**
 > System waits forever for `/dev/mapper/root`

**Steps to reproduce the problem**

My disk is encrypted using cryptsetup. `/dev/mapper/root` contains a btrfs filesystem, which contains the subvolume which is mounted on my / and others.
Within dracut, everything works as expected, but after the switch root, when systemd tries to mount all the things from fstab, it just waits forever for `/dev/mapper/root` to be ready.
Passing `systemd.unit=emergency.target` and then running `systemctl start default.target` from the emergency shell sometimes works around this (`/dev/mapper/root` exists).

Rolling back to the previous merge (278939be7a8eb07c696b3601cb63522d252b7291) fixes the issue, master is broken since 4874191d9d766093f92059cbc0724bf2a25f6460 which is the merge commit of #10787

### PR DESCRIPTION
**Pull request by yuwata**
~~I hope this fixes #10854.~~

cc @Keruspe.

### CODE REVIEW

File: src/libudev/libudev-list-internal.h [hunk 1]
```diff
@@ -3,22 +3,17 @@
...
-        struct udev *udev;
-        struct udev_list_node node;
-        struct udev_list_entry **entries;
-        unsigned entries_cur;
-        unsigned entries_max;
+        Hashmap *hashmap;
         bool unique;
+
+        LIST_HEAD(struct udev_list_entry, entries);
 };
```
Comment (dnicolodi) [inline src/libudev/libudev-list-internal.h:16]: IIUC the list here is used only as support storage in `udev_hwdb_get_properties_list_entry()` to be able to return an  ordered list. Wouldn't it be better to use an `OrderedHashmap` instead? The implementation currently in the codebase only allows to store elements in insertion order, but it should not be hard to add a `sort` method to it.

Comment (yuwata) [reply src/libudev/libudev-list-internal.h:16]: If I understand correctly, libudev-list sorts entries with its name, and optionally drops duplicated entries which have same name. So, I re-implement this by using 'Hashmap' and 'qsort'. After sorting, list structure is initialized.

Note that `OrderedHashmap` does not sort entries. It keeps the insertion order.

**Comment by yuwata** [comment]
As #10854 is not caused by udev-list, I've dropped most changes. PTAL.

File: src/libudev/libudev-device.c [hunk 2]
```diff
@@ -156,11 +148,10 @@ _public_ const char *udev_device_get_subsystem(struct udev_device *udev_device)
...
 
         r = sd_device_get_subsystem(udev_device->device, &subsystem);
-        if (r < 0) {
-                errno = -r;
-                return NULL;
-        } else if (!subsystem)
-                errno = ENODATA;
+        if (r < 0)
+                return_with_errno(NULL, r);
+        if (!subsystem)
```
Comment (keszybz) [inline src/libudev/libudev-device.c:153]: `sd_device_get_subsystem` returns `-ENOENT` if subsystem is not set, so this check can be dropped.

**Comment by keszybz** [comment]
```
M[K[   [0;31m*[0;1;31m*[0m[0;31m*[0m] (1 of 2) A start job is running for?-UEFI.device (1min 29s / 1min 30s)
M[K[[0;1;31m TIME [0m] Timed out waiting for device dev-disk-by\x2dlabel-UEFI.device.
[K[[0;1;33mDEPEND[0m] Dependency failed for /boot/efi.
[[0;1;33mDEPEND[0m] Dependency failed for Local File Systems.
```
I have seen this before, so it doesn't seem to be related.

**Pull request merged**